contracts/token/ERC1155/extensions/ERC1155URIStorage.sol

```diff
@@ -0,0 +1,50 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts v4.4.1 (token/ERC1155/extensions/ERC1155URIStorage.sol)
+
+pragma solidity ^0.8.0;
+
+import "../../../utils/Strings.sol";
+import "../ERC1155.sol";
+
+/**
+ * @dev ERC1155 token with storage based token URI management.
+ * Inspired by the ERC721URIStorage extension
+ */
+abstract contract ERC1155URIStorage is ERC1155 {
+    using Strings for uint256;
+
+    // Optional mapping for token URIs
+    mapping(uint256 => string) private _tokenURIs;
+
+    /**
+     * @dev {IERC721Metadata-tokenURI} token URIs for ERC1155 tokens.
+     */
+    function tokenURI(uint256 tokenId) public view virtual returns (string memory) {
```

Comment:
`tokenURI` is not part of ERC1155. 

IMO you should not define a new function but override the existing `uri` function.

this means using `string memory base = super.uri(tokenId);`

Reply:
If I did that, how would I get the base uri (that's returned by the current `uri` method) though?

With the current implementation we can support URIs of the type

1. `https://token.com/1` as well as
2. `https://token.com/one` as well as
3. `ipfs://someIpfsHash`

After overwriting the `uri` method with the new functionality we could only properly support 1 & 3, but not 2, right?
I mean, technically one could still set the "base" uri but not separately request it which seems to me quite inconvenient... What do you think?

Reply:
First things first. The ERC1155 interface documents the `uri(uint256)` method, but not `tokenURI` ... it the the first that is going to be used/called by all frontend/indexing services ... so that is what should be used. `tokenURI` is not part of this standard, so we just should expect anyone to use it.

That being said, my proposal is to do 

```
    function uri(uitn256 tokenId) public view virtual override returns (string memory) {
        string memory _tokenURI = _tokenURIs[tokenId];
        string memory base = super.uri(tokenId);

        // If there is no base URI, return the token URI.
        if (bytes(base).length == 0) {
            return _tokenURI;
        }
        // If both are set, concatenate the base URI and tokenURI (via abi.encodePacked).
        if (bytes(_tokenURI).length > 0) {
            return string(abi.encodePacked(base, _tokenURI));
        }

        // If there is a base URI set but no tokenURI, return the base URI.
        return base;
    }
```

This leaves multiple usecases:

- both `ERC1155._uri` and `ERC1155URIStorage._tokenURIs[tokenId]` are empty:
  → the result is empty
- `ERC1155._uri` is not empty and `ERC1155URIStorage._tokenURIs[tokenId]` is empty:
  → the result is `ERC1155._uri` (transparent)
- `ERC1155._uri` is empty and `ERC1155URIStorage._tokenURIs[tokenId]` is not empty:
  → the result is `ERC1155URIStorage._tokenURIs[tokenId]` (we set a token specific uri)
- both `ERC1155._uri` and `ERC1155URIStorage._tokenURIs[tokenId]` are not empty:
  → the result is the concatenation of both
  
As for your examples:

- `https://token.com/1`, `https://token.com/2`, `https://token.com/3`
→ you should not use `ERC1155URIStorage`, you should set the base uri to `https://token.com/` and override `uri()` to do 
```string.concat(super.uri, String.toString(tokenId))```
  
- `https://token.com/one`, `https://token.com/two`, `https://token.com/tree`
→ you should use  `ERC1155URIStorage`, set the base uri to `https://token.com/` and use the URIStorage feature to store "one", "two", "tree"

- `ipfs://someIpfsHash`, `ipfs://someotherIpfsHash`, ...
→ you could use `ERC1155URIStorage`, set the base uri to `ipfs://` and use URIStorage to manage the hashes

Reply:
@Amxx Fair enough, thanks for elaborating in such detail and adequacy. 👍

I guess the only case we didn't yet cover is the usage of a combo of different protocols, such as `https://` and `ipfs://`, e.g. `https://token.com/abc123` and `ipfs://someIpfsHash`.
Here we could just leave the base `ERC1155._uri` empty but if anybody were to change it at some point it would invalidate all the existing token uris. That being said, that's a limitation I can live with.

I'm going to update the code according to your proposal.

Reply:
@Amxx Do you know how to resolve [this error](https://github.com/OpenZeppelin/openzeppelin-contracts/runs/5295895376?check_suite_focus=true#step:7:13)? What I tried (not committed) was:

```solidity
// file: ERC1155URIStorageMock.sol
contract ERC1155URIStorageMock is ERC1155Mock, ERC1155URIStorage {
    constructor(string memory uri) ERC1155Mock(uri) {}

    function uri(uint256 tokenId) public view virtual override returns (string memory) {
        ERC1155URIStorage.uri(tokenId);
    }

    function setTokenURI(uint256 tokenId, string memory _tokenURI) public {
        _setTokenURI(tokenId, _tokenURI);
    }
}
```

However, that'd fail at `override` with:

> Function needs to specify overridden contracts "ERC1155" and "ERC1155URIStorage".